Add Bluesky backfeed support

JoelOtter opened the pull request:
This is the bridgy portion of the Bluesky work. It's rather simple compared to the Granary work, which this depends on along with the oauth-dropins work. 

Most significantly it adds the app password submission page, which currently looks like this:

![](https://user-images.githubusercontent.com/2765356/253829980-cac797bb-42db-4817-a7a5-c6386d542abe.png)

Tests to come.

snarfed commented:
Argh, just realized I never sent my review comments. Sorry for the delay!

snarfed commented:
(this is re: #1453)

JoelOtter commented on bluesky.py:
```diff
@@ -0,0 +1,81 @@
+    return self.gr_source.user_url(self.name)
+
+  def format_for_source_url(self, key):
+    """Bluesky keys (AT URIs) contain slashes, so must be double-encoded."""
```
This is due to an issue in Flask (via WSGI) wherein URL-encoded slashes are treated as slashes, which breaks routing https://github.com/pallets/flask/issues/900

snarfed replied:
Woowww that's unfortunate. Could you add that explanation and link to the docstring, including that we don't have to decode them correspondingly?

JoelOtter commented on handlers.py:
```diff
@@ -76,7 +77,7 @@ class Item(View):
   """
   source = None
 
-  VALID_ID = re.compile(r'^[\w.+:@=<>-]+$')
```
Everything should be URL encoded so there shouldn't be colons - where would they come from?

snarfed replied:
We don't really try hard to guarantee URL-encoding. Eg here's the code that generates webmention source URLs:

https://github.com/snarfed/bridgy/blob/1d1fc440a504acb53333121215f00da0ce9af466/tasks.py#L782-L798

URL encoding is nice, but I'm not that much of a stickler when it's not necessary for correctness. I sometimes like leaving in some characters like `:` when possible, even if only for aesthetics. As an example, when Bridgy supported Facebook, many of its ids included colons: https://github.com/snarfed/granary/blob/4f13ce738fabfdda11d59f96a3f6878df2233390/granary/facebook.py#L2427-L2447

snarfed replied:
Ah, I see your `format_for_source_url` change. OK! We may still get colons in other silos' ids, but at least not Bluesky's.

JoelOtter commented:
@snarfed Think I've addressed all comments - just needs tests added and the Granary change merged for it to pass on CI. I'm having a couple of tests fail locally and not sure why, would you be able to offer guidance?

```
======================================================================
FAIL: test_social_user_page_mf2 (tests.test_pages.PagesTest)
Check the custom mf2 we render on social user pages.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/joel/Code/bridgy/local/lib/python3.9/site-packages/mox3/mox.py", line 2139, in new_method
    func(self, *args, **kwargs)
  File "/Users/joel/Code/bridgy/tests/test_pages.py", line 328, in test_social_user_page_mf2
    self.assertEqual(200, response.status_code)
AssertionError: 200 != 500

======================================================================
FAIL: test_user_page_escapes_html_chars (tests.test_pages.PagesTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/joel/Code/bridgy/local/lib/python3.9/site-packages/mox3/mox.py", line 2139, in new_method
    func(self, *args, **kwargs)
  File "/Users/joel/Code/bridgy/tests/test_pages.py", line 418, in test_user_page_escapes_html_chars
    self.assertEqual(200, resp.status_code)
AssertionError: 200 != 500
```

snarfed commented:
Awesome! I'll look tomorrow morning. For the tests, try running just one of them with `-v`, eg `python -m unittest -v -k test_social_user_page_mf2`, you should see the stack trace where it's crashing along with full logging.

JoelOtter commented:
> Awesome! I'll look tomorrow morning. For the tests, try running just one of them with `-v`, eg `python -m unittest -v -k test_social_user_page_mf2`, you should see the stack trace where it's crashing along with full logging.

This worked great, thanks!

snarfed commented:
This and the granary change are both looking great! You're right, the last remaining thing I'd probably look for is new tests. Not necessarily the most fun part, but I pretty much always find more bugs when I write them. Most new tests would probably go into granary; some here too, but mostly just logic, eg the URL-encoding/decoding in `handlers.py` and `PropagateResponse.source_url`, no need to worry too much about trying to test UI stuff. Feel free to copy and paste liberally from existing tests, obviously!

JoelOtter commented:
@snarfed How's this for tests? More needed? I kind of followed the Mastodon tests

JoelOtter commented:
CI errors appear to be hitting that old version of lexrpc again 🙃

snarfed commented:
> CI errors appear to be hitting that old version of lexrpc again 🙃

Oops true, sorry about that! Hadn't fixed that in this repo yet. I've done that now, feel free to rebase and try again.

JoelOtter commented:
Hmm, seem to be getting the same thing:

`TypeError: Client.__init__() missing 1 required positional argument: 'lexicons'`

snarfed commented:
Huh, true, that's odd. Looking.

snarfed commented:
Ah, it wasn't installing lexrpc from github. Fixed that! I'll look at the PR itself now...

snarfed commented on bluesky.py:
```diff
@@ -0,0 +1,90 @@
+    gr_source = gr_bluesky.Bluesky(user.get('handle'), app_password=app_password)
+    actor = gr_source.user_to_actor(user)
+    return Bluesky(id=auth_entity.key_id(),
+                   username=auth_entity.key_id(),
```
this should probably be handle, right?

JoelOtter replied:
Are you saying the value for username should be the Bluesky handle rather than the DID, or I should be passing a kwarg `handle` into the constructor?

JoelOtter replied:
I think I'm being stupid here and you definitely mean the former, will change

snarfed replied:
yes! 😁

snarfed commented on bluesky.py:
```diff
@@ -0,0 +1,90 @@
+    return Bluesky(id=auth_entity.key_id(),
+                   username=auth_entity.key_id(),
+                   auth_entity=auth_entity.key,
+                   name=user.get('handle'),
```
and then this should be displayName?

JoelOtter replied:
It could be but I think that would actually be inconsistent with other silos? Bridgy current seems to use username rather than display name for this in Mastodon and Twitter (RIP):

<img width="469" alt="image" src="https://github.com/snarfed/bridgy/assets/2765356/a4c3ef42-9fad-4f40-97e8-3256f74a5fb5">

JoelOtter replied:
Hmm, unless this is something that changed recently? Looking at the source code it appears to be as you say, but that isn't reflected in my actual brid.gy account

snarfed replied:
right, Bridgy shows silo-specific usernames/handles in the UI, but those come from `label_name()`, not from the `name` property. I think we want `name` to be the human-readable name.

https://github.com/snarfed/bridgy/blob/33b4a717c0c0686ecc308d15b55836e23801b8e7/templates/profile_link.html#L12

https://github.com/snarfed/bridgy/blob/33b4a717c0c0686ecc308d15b55836e23801b8e7/models.py#L143

Checking in a few other silos...yup, `name` should be human name.

JoelOtter replied:
OK I'm a bit baffled by this, running it locally gives the same result, but the name field is clearly my display name, it just doesn't get used in the display, but in Bluesky it does!

<img width="540" alt="image" src="https://github.com/snarfed/bridgy/assets/2765356/08a20d0f-39b3-40d8-8fa1-f4e16a769236">

I guess the problem is somewhere else?

JoelOtter replied:
Ah, so I should change the implementation of `label_name()`?

snarfed replied:
Ahh sorry, the missing piece is that `Bluesky` itself should override `label_name()` with whatever we want there. Other examples:

https://github.com/snarfed/bridgy/blob/main/mastodon.py#L119-L121

https://github.com/snarfed/bridgy/blob/33b4a717c0c0686ecc308d15b55836e23801b8e7/github.py#L74-L76

so yeah I think we want to do that and return handle.

snarfed replied:
Sorry for the confusion!

JoelOtter replied:
I wish GitHub had auto-refresh 😆 Will get that changed!

snarfed commented on bluesky.py:
```diff
@@ -0,0 +1,90 @@
+  OAUTH_START = oauth_bluesky.Start
+  AUTH_MODEL = oauth_bluesky.BlueskyAuth
+  URL_CANONICALIZER = util.UrlCanonicalizer(
+          domain=GR_CLASS.DOMAIN,
```
this makes me think we need to support multiple domains in UrlCanonicalizer, so we can canonicalize staging.bsky.app => bsky.app. Obviously not needed in this PR though!

JoelOtter replied:
Ugh yeah federation/multiple PDSs is just not something I've thought about yet, likely to be a headache but hoping we can use a lot of the same stuff Mastodon does?

snarfed replied:
Oh this isn't federation, that's just the main official PSS's old hostname, there are still syndication links to it out there in the wild, eg on my site 😁

Agreed though, let's definitely punt on federation itself for now!

snarfed commented on bluesky.py:
```diff
@@ -0,0 +1,90 @@
+
+  @classmethod
+  def button_html(cls, feature):
+    """Override oauth-dropins's button_html() to not show the instance text box."""
```
minor nit: update docstring? no instance text box, right?

snarfed reviewed (commented):
This is looking pretty ready! Couple minor nits, but nothing big. Ready to merge after those and CI pass!

snarfed commented:
Oh, one last small thing, could you center the handle + app password text boxes on the login page?

JoelOtter commented:
Should be all addressed now!

snarfed commented:
Woo, looks great! Thanks for your patience. Merging and deploying! Can't wait to try it in prod!

Next step, docs? Not much, only a few sections on https://brid.gy/about and in the README that call out specific silos. Then maybe a blog post? You should get to do the honors!

snarfed commented:
Background service isn't starting up ok. 😕 Not sure why, I'm guessing not related to this PR, maybe a bad new dependency that we upgraded since deploying last a couple weeks ago. Looking now...

snarfed commented:
Startup command is `gunicorn --workers 1 --threads 10 --timeout 600 -b :$PORT background:app`, that runs fine locally for me on macOS. Hrm.

snarfed commented:
Same build config for both services' containers, and the default (ie frontend) service is up and serving fine. Can't tell why background isn't. And I'm not seeing any useful logs, it just closes the `/_ah/start` request's connection after 5s, which makes the watchdog terminate it. Hrm.

snarfed commented:
Rolling back for now, still investigating.

snarfed commented:
Aha, `background.yaml` is still on `runtime: python39`, missed upgrading that to 3.11 a while back. Fingers crossed that fixes it...

snarfed commented:
It's alive, it's aliiiive!

JoelOtter commented:
> Next step, docs? Not much, only a few sections on https://brid.gy/about and in the README that call out specific silos. Then maybe a blog post? You should get to do the honors!

Very happy to write some docs. :) What would a blog post look like, do you mean on my own blog?

snarfed commented:
Sure, at least if you want! Just announcing the launch, and maybe syndicated to https://news.indieweb.org/en . Totally optional, of course.

JoelOtter commented:
Cool! I'm guessing the indie web way ™️ would be to write it on my own site, do a `u-syndication` to news.indieweb, do a `u-mention-of` to yourself, then send the webmentions accordingly?

snarfed commented:
`u-syndication` to IndieNews, yes! `u-mention-of` to...me? Not necessary but OK if you want! Except I don't think `u-mention-of` exists, mentions are just links, no need to mark up with any specific mf2.
